modal/_utils/shell_utils.py

```diff
@@ -0,0 +1,112 @@
+# Copyright Modal Labs 2024
+import asyncio
+import contextlib
+import errno
+import os
+import select
+import sys
+from typing import Callable, Coroutine, Optional
+
+import rich.status
+
+from modal._pty import raw_terminal, set_nonblocking
+from modal.exception import ExecutionError, InteractiveTimeoutError
+
+from .async_utils import TaskContext, asyncify
+
+
+def write_to_fd(fd: int, data: bytes):
+    loop = asyncio.get_event_loop()
+    future = loop.create_future()
+
+    def try_write():
+        try:
+            nbytes = os.write(fd, data)
+            loop.remove_writer(fd)
+            future.set_result(nbytes)
+        except OSError as e:
+            if e.errno != errno.EAGAIN:
+                future.set_exception(e)
+                raise
+
+    loop.add_writer(fd, try_write)
+    return future
+
+
+@contextlib.asynccontextmanager
+async def _stream_stdin(handle_input: Callable[[bytes, int], Coroutine], use_raw_terminal=False):
```

Comment:
this method is moved and renamed from `container_exec::handle_exec_input` because it is now also used by `Sandbox` for `shell`